File: app/src/main/java/com/kotobyte/search/page/SearchResultsAdapter.java
package com.kotobyte.search.page;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kotobyte.R;
import com.kotobyte.databinding.ViewWordItemCollapsedBinding;
import com.kotobyte.databinding.ViewWordItemExpandedBinding;
import com.kotobyte.models.Kanji;
import com.kotobyte.models.Word;

import java.util.List;


class SearchResultsAdapter extends RecyclerView.Adapter {

    private static final int COLLAPSED_ITEM = 1;
    private static final int EXPANDED_ITEM = 2;
    private static final int NONE = -1;

    private Context mContext;
    private Listener mListener;

    private int mExpandedItemPosition = NONE;
    private List<Word> mWordSearchResults;
    private SparseArray<KanjiSearchResultsAdapter> mKanjiSearchResultsAdapters;

    private WordLiteralsTextGenerator mLiteralsTextGenerator;
    private WordSensesTextGenerator mSensesTextGenerator;

    SearchResultsAdapter(Context context, Listener listener, List<Word> wordSearchResults) {
        mContext = context;
        mListener = listener;

        mWordSearchResults = wordSearchResults;
        mKanjiSearchResultsAdapters = new SparseArray<>();

        mLiteralsTextGenerator = new WordLiteralsTextGenerator(context, mWordSearchResults);
        mSensesTextGenerator = new WordSensesTextGenerator(context, mWordSearchResults);

        setHasStableIds(true);
    }

    void setKanjiSearchResults(int position, List<Kanji> kanjiSearchResults) {

        mKanjiSearchResultsAdapters.put(position,
                new KanjiSearchResultsAdapter(mContext, mKanjiAdapterListener, kanjiSearchResults));

        notifyItemChanged(position);
    }

    @Override
    public int getItemCount() {
        return mWordSearchResults.size();
    }

    @Override
    public long getItemId(int position) {
        return mWordSearchResults.get(position).getID();
    }

    @Override
    public int getItemViewType(int position) {
        return mExpandedItemPosition == position ? EXPANDED_ITEM : COLLAPSED_ITEM;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        if (viewType == EXPANDED_ITEM) {
            ViewWordItemExpandedBinding binding = DataBindingUtil.inflate(
                    LayoutInflater.from(parent.getContext()), R.layout.view_word_item_expanded, parent, false);

            RecyclerView.ViewHolder viewHolder = new ExpandedViewHolder(binding);
            binding.wordItem.wordContainer.setOnClickListener(new OnWordClickListener(viewHolder));

            return viewHolder;

        } else {
            ViewWordItemCollapsedBinding binding = DataBindingUtil.inflate(
                    LayoutInflater.from(parent.getContext()), R.layout.view_word_item_collapsed, parent, false);

            RecyclerView.ViewHolder viewHolder = new CollapsedViewHolder(binding);
            binding.wordItem.wordContainer.setOnClickListener(new OnWordClickListener(viewHolder));

            return viewHolder;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {

        if (getItemViewType(position) == EXPANDED_ITEM) {
            ExpandedViewHolder viewHolder = (ExpandedViewHolder) holder;

            viewHolder.mBinding.wordItem.literalsTextView.setText(mLiteralsTextGenerator.getSpannableString(position));
            viewHolder.mBinding.wordItem.sensesTextView.setText(mSensesTextGenerator.getSpannableString(position));

            viewHolder.mBinding.progressBar.setVisibility(View.GONE);
            viewHolder.mBinding.kanjiListView.setVisibility(View.GONE);

            KanjiSearchResultsAdapter kanjiSearchResultsAdapter = mKanjiSearchResultsAdapters.get(position);

            if (kanjiSearchResultsAdapter == null) {
                viewHolder.mBinding.progressBar.setVisibility(View.VISIBLE);

            } else {
                viewHolder.mBinding.kanjiListView.setVisibility(View.VISIBLE);
                viewHolder.mBinding.kanjiListView.setAdapter(kanjiSearchResultsAdapter);
            }

        } else {
            CollapsedViewHolder viewHolder = (CollapsedViewHolder) holder;

            viewHolder.mBinding.wordItem.literalsTextView.setText(mLiteralsTextGenerator.getSpannableString(position));
            viewHolder.mBinding.wordItem.sensesTextView.setText(mSensesTextGenerator.getSpannableString(position));
        }
    }

    private KanjiSearchResultsAdapter.Listener mKanjiAdapterListener = new KanjiSearchResultsAdapter.Listener() {

        @Override
        public void onClickKanji(int position, Kanji kanji) {
            mListener.onRequestDetailForKanji(kanji);
        }
    };

    interface Listener {
        void onRequestKanjiListForWord(int position, Word word);
        void onRequestDetailForKanji(Kanji kanji);
    }

    private static class ExpandedViewHolder extends RecyclerView.ViewHolder {

        private ViewWordItemExpandedBinding mBinding;

        ExpandedViewHolder(ViewWordItemExpandedBinding binding) {
            super(binding.getRoot());

            mBinding = binding;
        }
    }

    private static class CollapsedViewHolder extends RecyclerView.ViewHolder {

        private ViewWordItemCollapsedBinding mBinding;

        CollapsedViewHolder(ViewWordItemCollapsedBinding binding) {
            super(binding.getRoot());

            mBinding = binding;
        }
    }

    private class OnWordClickListener implements View.OnClickListener {

        private RecyclerView.ViewHolder mViewHolder;

        OnWordClickListener(RecyclerView.ViewHolder viewHolder) {
            mViewHolder = viewHolder;
        }

        @Override
        public void onClick(View v) {
            int position = mViewHolder.getAdapterPosition();

            if (mExpandedItemPosition == position) {
                mExpandedItemPosition = NONE;

                notifyItemChanged(position);

            } else {
                int lastExpandedItemPosition = mExpandedItemPosition;

                mExpandedItemPosition = position;

                if (lastExpandedItemPosition != NONE) {
                    notifyItemChanged(lastExpandedItemPosition);
                }

                notifyItemChanged(mExpandedItemPosition);

                if (mKanjiSearchResultsAdapters.get(position) == null) {
                    mListener.onRequestKanjiListForWord(position, mWordSearchResults.get(position));
                }
            }
        }
    }
 }

File: app/src/main/java/com/kotobyte/search/SearchResultsAdapter.kt
package com.kotobyte.search

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.text.SpannableString
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kotobyte.R
import com.kotobyte.databinding.ViewWordItemCollapsedBinding
import com.kotobyte.databinding.ViewWordItemExpandedBinding
import com.kotobyte.models.Kanji
import com.kotobyte.models.Word


internal class SearchResultsAdapter(
        private val context: Context,
        private val listener: Listener,
        private val words: List<Word>

) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var expandedItemPosition = NONE
    private val kanjiSearchResultsAdapter = SparseArray<KanjiSearchResultsAdapter>()

    private val literalsTextGenerator = WordLiteralsTextGenerator(context)
    private val sensesTextGenerator = WordSensesTextGenerator(context)

    private val literalsTexts = SparseArray<SpannableString>(words.size)
    private val sensesTexts = SparseArray<SpannableString>(words.size)

    private val kanjiAdapterListener = object : KanjiSearchResultsAdapter.Listener {

        override fun onClickKanji(position: Int, kanji: Kanji) =
                listener.onRequestDetailForKanji(kanji)
    }

    init {
        setHasStableIds(true)
    }

    fun setKanjiSearchResults(position: Int, kanjiSearchResults: List<Kanji>) {

        kanjiSearchResultsAdapter.put(position,
                KanjiSearchResultsAdapter(context, kanjiAdapterListener, kanjiSearchResults))

        notifyItemChanged(position)
    }

    override fun getItemCount(): Int = words.size

    override fun getItemId(position: Int): Long = words[position].ID

    override fun getItemViewType(position: Int): Int =
            if (expandedItemPosition == position) EXPANDED_ITEM else COLLAPSED_ITEM

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        if (viewType == EXPANDED_ITEM) {

            val binding = DataBindingUtil.inflate<ViewWordItemExpandedBinding>(
                    LayoutInflater.from(parent.context), R.layout.view_word_item_expanded, parent, false)

            return ExpandedViewHolder(binding).apply {
                binding.wordItem.wordContainer.setOnClickListener(OnWordClickListener(this))
            }

        } else {
            val binding = DataBindingUtil.inflate<ViewWordItemCollapsedBinding>(
                    LayoutInflater.from(parent.context), R.layout.view_word_item_collapsed, parent, false)

            return CollapsedViewHolder(binding).apply {
                binding.wordItem.wordContainer.setOnClickListener(OnWordClickListener(this))
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        val literalsText = literalsTexts.get(position) ?:
                literalsTextGenerator.createFrom(words[position]).also { literalsTexts.put(position, it) }

        val sensesText = sensesTexts.get(position) ?:
                sensesTextGenerator.createFrom(words[position]).also { sensesTexts.put(position, it) }

        if (getItemViewType(position) == EXPANDED_ITEM) {
            val viewHolder = holder as ExpandedViewHolder

            viewHolder.binding.wordItem.literalsTextView.text = literalsText
            viewHolder.binding.wordItem.sensesTextView.text = sensesText

            viewHolder.binding.progressBar.visibility = View.GONE
            viewHolder.binding.kanjiListView.visibility = View.GONE

            val kanjiSearchResultsAdapter = kanjiSearchResultsAdapter.get(position)

            if (kanjiSearchResultsAdapter == null) {
                viewHolder.binding.progressBar.visibility = View.VISIBLE

            } else {
                viewHolder.binding.kanjiListView.visibility = View.VISIBLE
                viewHolder.binding.kanjiListView.adapter = kanjiSearchResultsAdapter
            }

        } else {
            val viewHolder = holder as CollapsedViewHolder

            viewHolder.binding.wordItem.literalsTextView.text = literalsText
            viewHolder.binding.wordItem.sensesTextView.text = sensesText
        }
    }

    internal interface Listener {
        fun onRequestKanjiListForWord(position: Int, word: Word)
        fun onRequestDetailForKanji(kanji: Kanji)
    }

    private class ExpandedViewHolder(val binding: ViewWordItemExpandedBinding) : RecyclerView.ViewHolder(binding.root)
    private class CollapsedViewHolder(val binding: ViewWordItemCollapsedBinding) : RecyclerView.ViewHolder(binding.root)

    private inner class OnWordClickListener internal constructor(private val mViewHolder: RecyclerView.ViewHolder) : View.OnClickListener {

        override fun onClick(v: View) {
            val position = mViewHolder.adapterPosition

            if (expandedItemPosition == position) {
                expandedItemPosition = NONE

                notifyItemChanged(position)

            } else {
                val lastExpandedItemPosition = expandedItemPosition

                expandedItemPosition = position

                if (lastExpandedItemPosition != NONE) {
                    notifyItemChanged(lastExpandedItemPosition)
                }

                notifyItemChanged(expandedItemPosition)

                if (kanjiSearchResultsAdapter.get(position) == null) {
                    listener.onRequestKanjiListForWord(position, words[position])
                }
            }
        }
    }

    companion object {

        private val COLLAPSED_ITEM = 1
        private val EXPANDED_ITEM = 2
        private val NONE = -1
    }
}
